formula/src/main/java/com/instacart/formula/Effect.kt

```diff
@@ -0,0 +1,47 @@
+package com.instacart.formula
+
+/**
+ * Effect is a function returned within [Transition.Result] which will be executed by
+ * [FormulaRuntime] usually after [Formula] state changes are applied. This function is
+ * the place to call listeners, log events, trigger network requests or database writes,
+ * and etc. Instead of executing effects within [Transition.toResult], deferring execution
+ * allows us to ensure that [Formula] is always in the correct state in case effects
+ * trigger a new event.
+ */
+data class Effect(
+    val type: Type,
+    val executable: () -> Unit,
+) {
+    constructor(executable: () -> Unit): this(Main, executable)
+
+    /**
+     * An effect type which defines the execution timing and threading of the effect. Take
+     * a look at [Unconfined], [Main], [Background].
+     */
+    sealed class Type
+
+    /**
+     * Unconfined effect type indicates that the effect can run on any thread and will be
+     * executed after all state changes are processed.
+     */
+    data object Unconfined : Type()
+
+    /**
+     * Immediate effect type indicates that the effect can run on any thread and can be executed
+     * immediately before the state changes are processed and formula is re-evaluated. This
+     * is a useful optimization for inlining multiple state changes as part of a single
+     * re-evaluation.
+     */
+    // TODO: this will be added in a future.
+//    data object Immediate: Type()
+
+    /**
+     * Will schedule side-effect to be executed on the main thread.
```

Comment:
Not directly related to this PR, b/c we already started to articulate "main thread" terminology in other places - should this "main-thread" limitations be exclusive to `formula-android` only? If we claim `formula` to be Android-agnostic and JVM-friendly or even kotlin-multiplatform-friendly, the concept of "only main thread is allowed to touch UI" is irrelevant there.

Reply:
`Effect.Main` and `Effect.Background` are metadata for `Plugin` to decide how to interpret. If a platform doesn't support `Main`, it doesn't have to do anything. I could have named it `Effect.Confined`, but thought `Effect.Main` seems clearer though less generic.